go/vt/vtgate/tx_conn.go

```diff
@@ -190,114 +197,118 @@ func (txc *TxConn) commit2PC(ctx context.Context, session *SafeSession) error {
-				return err
+	txPhase = COMMIT2PC_PREPARE
+	prepareAction := func(ctx context.Context, s *vtgatepb.Session_ShardSession, logging *executeLogger) error {
+		if DebugTwoPc { // Test code to simulate a failure during RM prepare
+			if terr := checkTestFailure(ctx, "RMPrepare_-40_FailNow", s.Target); terr != nil {
+				return terr
 			}
 		}
 		return txc.tabletGateway.Prepare(ctx, s.Target, s.TransactionId, dtid)
-	})
-	if err != nil {
-		// TODO(sougou): Perform a more fine-grained cleanup
-		// including unprepared transactions.
-		if resumeErr := txc.rollbackTx(ctx, dtid, mmShard, session.ShardSessions[1:], session.logging); resumeErr != nil {
-			log.Warningf("Rollback failed after Prepare failure: %v", resumeErr)
-		}
-		// Return the original error even if the previous operation fails.
+	}
+	if err = txc.runSessions(ctx, rmShards, session.logging, prepareAction); err != nil {
 		return err
 	}
 
-	if DebugTwoPc {
-		// Test code to simulate a failure after RM prepare
-		if failNow, err := checkTestFailure(callerid.EffectiveCallerIDFromContext(ctx), "RMPrepared_FailNow", nil); failNow {
-			return err
+	if DebugTwoPc { // Test code to simulate a failure after RM prepare
+		if terr := checkTestFailure(ctx, "RMPrepared_FailNow", nil); terr != nil {
+			return terr
 		}
 	}
 
+	txPhase = COMMIT2PC_STARTCOMMIT
 	err = txc.tabletGateway.StartCommit(ctx, mmShard.Target, mmShard.TransactionId, dtid)
 	if err != nil {
 		return err
 	}
 
-	if DebugTwoPc {
-		// Test code to simulate a failure after MM commit
-		if failNow, err := checkTestFailure(callerid.EffectiveCallerIDFromContext(ctx), "MMCommitted_FailNow", nil); failNow {
-			return err
+	if DebugTwoPc { // Test code to simulate a failure after MM commit
+		if terr := checkTestFailure(ctx, "MMCommitted_FailNow", nil); terr != nil {
+			return terr
 		}
 	}
 
-	err = txc.runSessions(ctx, session.ShardSessions[1:], session.logging, func(ctx context.Context, s *vtgatepb.Session_ShardSession, logging *executeLogger) error {
-		if DebugTwoPc {
-			// Test code to simulate a failure during RM prepare
-			if failNow, err := checkTestFailure(callerid.EffectiveCallerIDFromContext(ctx), "RMCommit_-40_FailNow", s.Target); failNow {
-				return err
+	txPhase = COMMIT2PC_PREPARECOMMIT
+	prepareCommitAction := func(ctx context.Context, s *vtgatepb.Session_ShardSession, logging *executeLogger) error {
+		if DebugTwoPc { // Test code to simulate a failure during RM prepare
+			if terr := checkTestFailure(ctx, "RMCommit_-40_FailNow", s.Target); terr != nil {
+				return terr
 			}
 		}
 		return txc.tabletGateway.CommitPrepared(ctx, s.Target, dtid)
-	})
-	if err != nil {
+	}
+	if err = txc.runSessions(ctx, rmShards, session.logging, prepareCommitAction); err != nil {
 		return err
 	}
 
-	return txc.tabletGateway.ConcludeTransaction(ctx, mmShard.Target, dtid)
+	// At this point, application can continue forward.
```

Comment:
for my own understanding - could we in theory do the clean up in a background thread and answer the user that their commit has been accepted?

Reply:
In theory, yes, but it is better to call conclude now, than using goroutine or transaction resolver to conclude.